usr/src/man/man1/crontab.1
.\"
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for
.\" permission to reproduce portions of its copyrighted documentation.
.\" Original documentation from The Open Group can be obtained online at
.\" http://www.opengroup.org/bookstore/.
.\"
.\" The Institute of Electrical and Electronics Engineers and The Open
.\" Group, have given us permission to reprint portions of their
.\" documentation.
.\"
.\" In the following statement, the phrase ``this text'' refers to portions
.\" of the system documentation.
.\"
.\" Portions of this text are reprinted and reproduced in electronic form
.\" in the SunOS Reference Manual, from IEEE Std 1003.1, 2004 Edition,
.\" Standard for Information Technology -- Portable Operating System
.\" Interface (POSIX), The Open Group Base Specifications Issue 6,
.\" Copyright (C) 2001-2004 by the Institute of Electrical and Electronics
.\" Engineers, Inc and The Open Group.  In the event of any discrepancy
.\" between these versions and the original IEEE and The Open Group
.\" Standard, the original IEEE and The Open Group Standard is the referee
.\" document.  The original Standard can be obtained online at
.\" http://www.opengroup.org/unix/online.html.
.\"
.\" This notice shall appear on any product containing this material.
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\"
.\" Copyright 1989 AT&T
.\" Portions Copyright (c) 1992, X/Open Company Limited All Rights Reserved
.\" Copyright (c) 2009, Sun Microsystems, Inc. All Rights Reserved
.\" Copyright 2019 OmniOS Community Edition (OmniOSce) Association.
.\"
.TH CRONTAB 1 "Oct 22, 2019"
.SH NAME
crontab \- user crontab file
.SH SYNOPSIS
.LP
.nf
\fB/usr/bin/crontab\fR [\fIfilename\fR]
.fi

.LP
.nf
\fB/usr/bin/crontab\fR \fB-e\fR [\fIusername\fR]
.fi

.LP
.nf
\fB/usr/bin/crontab\fR \fB-l\fR [\fIusername\fR]
.fi

.LP
.nf
\fB/usr/bin/crontab\fR \fB-r\fR [\fIusername\fR]
.fi

.LP
.nf
\fB/usr/xpg4/bin/crontab\fR [\fIfilename\fR]
.fi

.LP
.nf
\fB/usr/xpg4/bin/crontab\fR \fB-e\fR [\fIusername\fR]
.fi

.LP
.nf
\fB/usr/xpg4/bin/crontab\fR \fB-l\fR [\fIusername\fR]
.fi

.LP
.nf
\fB/usr/xpg4/bin/crontab\fR \fB-r\fR [\fIusername\fR]
.fi

.LP
.nf
\fB/usr/xpg6/bin/crontab\fR [\fIfilename\fR]
.fi

.LP
.nf
\fB/usr/xpg6/bin/crontab\fR \fB-e\fR [\fIusername\fR]
.fi

.LP
.nf
\fB/usr/xpg6/bin/crontab\fR \fB-l\fR [\fIusername\fR]
.fi

.LP
.nf
\fB/usr/xpg6/bin/crontab\fR \fB-r\fR [\fIusername\fR]
.fi

.SH DESCRIPTION
.LP
The \fBcrontab\fR utility manages a user's access with \fBcron\fR (see
\fBcron\fR(1M)) by copying, creating, listing, and removing \fBcrontab\fR
files. If invoked without options, \fBcrontab\fR copies the specified file, or
the standard input if no file is specified, into a directory that holds all
users' crontabs.
.sp
.LP
If \fBcrontab\fR is invoked with \fIfilename\fR, this overwrites an existing
\fBcrontab\fR entry for the user that invokes it.
.SS "\fBcrontab\fR Access Control"
.LP
Users: Access to \fBcrontab\fR is allowed:
.RS +4
.TP
.ie t \(bu
.el o
if the user's name appears in \fB/etc/cron.d/cron.allow\fR.
.RE
.RS +4
.TP
.ie t \(bu
.el o
if \fB/etc/cron.d/cron.allow\fR does not exist and the user's name is not in
\fB/etc/cron.d/cron.deny\fR.
.RE
.sp
.LP
Users: Access to \fBcrontab\fR is denied:
.RS +4
.TP
.ie t \(bu
.el o
if \fB/etc/cron.d/cron.allow\fR exists and the user's name is not in it.
.RE
.RS +4
.TP
.ie t \(bu
.el o
if \fB/etc/cron.d/cron.allow\fR does not exist and user's name is in
\fB/etc/cron.d/cron.deny\fR.
.RE
.RS +4
.TP
.ie t \(bu
.el o
if neither file exists, only a user with the \fBsolaris.jobs.user\fR
authorization is allowed to submit a job.
.RE
.RS +4
.TP
.ie t \(bu
.el o
if Auditing is enabled, the user's shell is not audited and the user is
not the \fBcrontab\fR owner. This can occur if the user logs in by way of a
program, such as some versions of \fBSSH\fR, which does not set audit
parameters.
.RE
.sp
.LP
The rules for \fBallow\fR and \fBdeny\fR apply to \fBroot\fR only if the
\fBallow\fR/\fBdeny\fR files exist.
.sp
.LP
The \fBallow\fR/\fBdeny\fR files consist of one user name per line.
.SS "\fBcrontab\fR Entry Format"
.LP
A \fBcrontab\fR file consists of lines of six fields each. The fields are
separated by spaces or tabs. The first five are integer patterns that specify
the following:
.sp
.in +2
.nf
minute (0\(mi59),
hour (0\(mi23),
day of the month (1\(mi31),
month of the year (1\(mi12),
day of the week (0\(mi6 with 0=Sunday).
.fi
.in -2
.sp

.sp
.LP
Each of these patterns can be either an asterisk (meaning all legal values) or
a list of elements separated by commas. An element is either a number or two
numbers separated by a minus sign (meaning an inclusive range).
.LP
A range or asterisk can optionally be followed by a step value as
\fI/<number>\fR. For example, \fI2\(mi59/3\fR can be used in the minutes field
to specify every three minutes starting at 2 past the hour, or \fI*/2\fR in
the hours field means every two hours.
.LP
Time specified here is interpreted in the currently active timezone. At the top
of the crontab file this is the timezone which is set system-wide in
/etc/default/init. A user can add a line such as:
.sp
.in +2
.nf
TZ=\fItimezone\fR
.fi
.in -2
.sp

.sp
.LP
\&...and all subsequent entries will be interpreted using that timezone, until
a new \fBTZ=\fR\fItimezone\fR line is encountered. The specification of days
can be made by two fields (day of the month and day of the week). Both are
adhered to if specified as a list of elements. See \fBEXAMPLES\fR.
.sp
.LP
The sixth field of a line in a \fBcrontab\fR file is a string that is executed
by the shell at the specified times. A percent character in this field (unless
escaped by \fB\e\fR\|) is translated to a \fINEWLINE\fR character.
.sp
.LP
Only the first line (up to a \fB`\|%\|'\fR or end of line) of the command field
is executed by the shell. Other lines are made available to the command as
standard input. Any blank line or line beginning with a \fB`\|#\|'\fR is a
comment and is ignored.
.sp
.LP
The shell is invoked from your $HOME directory. As with $TZ, both $SHELL and
$HOME can be set by having a line such as:
.sp
.in +2
.nf
SHELL=/usr/bin/\fIsomeshell\fR
.fi
.in -2
.sp

.sp
.LP
\&...or:
.sp
.in +2
.nf
HOME=\fIsomedirectory\fR
.fi
.in -2
.sp

.sp
.LP
\&...which will take precedence for all the remaining entries in the
\fBcrontab\fR or until there is another \fBHOME\fR or \fBSHELL\fR entry. It is
invoked with an \fBarg0\fR of the basename of the $SHELL that is currently in
effect. A user who wants to have his \fB\&.profile\fR or equivalent file
executed must  explicitly do so in the \fBcrontab\fR file. \fBcron\fR supplies
a default environment for every shell, defining HOME, LOGNAME, SHELL, TZ, and
PATH. The default PATH for user \fBcron\fR jobs is \fB/usr/bin\fR; while root
\fBcron\fR jobs default to \fB/usr/sbin:/usr/bin\fR. The default PATH can be
set in \fB/etc/default/cron\fR (see \fBcron\fR(1M)). The TZ, HOME, and SHELL
environment variables are set to match those that are in effect in the
\fBcrontab\fR file at the time.
.sp
.LP
If you do not redirect the standard output and standard error of your commands,
any generated output or errors are mailed to you.
.SS "\fBcrontab\fR Environment Variables"
.LP
The following variables are supported:
.sp
.ne 2
.na
\fB\fBHOME\fR\fR
.ad
.sp .6
.RS 4n
Allows the user to choose and alternative directory for cron to change
directory to prior to running the command. For example:
.sp
.in +2
.nf
HOME=/var/tmp
.fi
.in -2
.sp

.RE

.sp
.ne 2
.na
\fB\fBSHELL\fR\fR
.ad
.sp .6
.RS 4n
The name of the shell to use to run subsequent commands. For example:
.sp
.in +2
.nf
SHELL=/usr/bin/ksh
.fi
.in -2
.sp

.RE

.sp
.ne 2
.na
\fB\fBTZ\fR\fR
.ad
.sp .6
.RS 4n
Allows the user to choose the timezone in which the \fBcron\fR entries are run.
This effects both the environment of the command that is run and the timing of
the entry. For example, to have your entries run using the timezone for
Iceland, use:
.sp
.in +2
.nf
TZ=Iceland
.fi
.in -2
.sp

.RE

.sp
.LP
Each of these variables affects all of the lines that follow it in the
\fBcrontab\fR file, until it is reset by a subsequent line resetting that
variable. Hence, it is possible to have multiple timezones supported within a
single \fBcrontab\fR file.
.sp
.LP
The lines that are not setting these environment variables are  the same as
crontab entries that conform to the UNIX standard and are described elsewhere
in this man page.
.SS "Setting \fBcron\fR Jobs Across Timezones"
.LP
The default timezone of the \fBcron\fR daemon sets the system-wide timezone for
\fBcron\fR entries. This, in turn, is by set by default system-wide using
\fB/etc/default/init\fR.
.sp
.LP
If some form of \fBdaylight savings\fR or \fBsummer/winter time\fR is in
effect, then jobs scheduled during the switchover period could be executed
once, twice, or not at all.
.SH OPTIONS
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB-e\fR\fR
.ad
.RS 6n
Edits a copy of the current user's \fBcrontab\fR file, or creates an empty file
to edit if \fBcrontab\fR does not exist. When editing is complete, the file is
installed as the user's \fBcrontab\fR file.
.sp
The environment variable \fBEDITOR\fR determines which editor is invoked with
the \fB-e\fR option. All \fBcrontab\fR jobs should be submitted using
\fBcrontab\fR. Do not add jobs by just editing the \fBcrontab\fR file, because
\fBcron\fR is not aware of changes made this way.
.sp
If all lines in the \fBcrontab\fR file are deleted, the old \fBcrontab\fR file
is restored. The correct way to delete all lines is to remove the \fBcrontab\fR
file using the \fB-r\fR option.
.sp
If \fIusername\fR is specified, the specified user's \fBcrontab\fR file is
edited, rather than the current user's \fBcrontab\fR file. This can only be
done by root or by a user with the \fBsolaris.jobs.admin\fR authorization.
.RE

.sp
.ne 2
.na
\fB\fB-l\fR\fR
.ad
.RS 6n
Lists the \fBcrontab\fR file for the invoking user. Only root or a user with
the \fBsolaris.jobs.admin\fR authorization can specify a username following the
\fB-l\fR option to list the \fBcrontab\fR file of the specified user.
.RE

.sp
.ne 2
.na
\fB\fB-r\fR\fR
.ad
.RS 6n
Removes a user's \fBcrontab\fR from the \fBcrontab\fR directory. Only root or a
user with the \fBsolaris.jobs.admin\fR authorization can specify a username
following the \fB-r\fR option to remove the \fBcrontab\fR file of the specified
user.
.RE

.SH EXAMPLES
.LP
\fBExample 1 \fRCleaning up Core Files
.sp
.LP
This example cleans up \fBcore\fR files every weekday morning at 3:15 am:

.sp
.in +2
.nf
\fB15 3 * * 1-5 find $HOME\fR \fB-name\fR\fBcore 2>/dev/null | xargs rm\fR \fB-f\fR
.fi
.in -2
.sp

.LP
\fBExample 2 \fRMailing a Birthday Greeting
.sp
.LP
This example mails a birthday greeting:

.sp
.in +2
.nf
\fB0 12 14 2 * mailx john%Happy Birthday!%Time for lunch.\fR
.fi
.in -2
.sp

.LP
\fBExample 3 \fRSpecifying Days of the Month and Week
.sp
.LP
This example runs a command on the first and fifteenth of each month, as well
as on every Monday:

.sp
.in +2
.nf
\fB0 0 1,15 * 1\fR
.fi
.in -2
.sp

.sp
.LP
To specify days by only one field, the other field should be set to *. For
example:

.sp
.in +2
.nf
\fB0 0 * * 1\fR
.fi
.in -2
.sp

.sp
.LP
would run a command only on Mondays.

.LP
\fBExample 4 \fRUsing step values:
.sp
.LP
This example runs a job every hour during the night and every 3 hours during
working hours.

.sp
.in +2
.nf
\fB0 0-7,8-18/3,19-23 * * *\fR
.fi
.in -2
.sp

.LP
and to run a job every 2 minutes, use:

.sp
.in +2
.nf
\fB*/2 * * * *\fR
.fi
.in -2
.sp

.LP
\fBExample 5 \fRUsing Environment Variables
.sp
.LP
The following entries take advantage of \fBcrontab\fR support for certain
environment variables.

.sp
.in +2
.nf
TZ=GMT
HOME=/local/home/user
SHELL=/usr/bin/ksh
0 0 * * * echo $(date) >        midnight.GMT
TZ=PST
0 0 * * * echo $(date) >        midnight.PST
TZ=EST
HOME=/local/home/myuser
SHELL=/bin/csh
.fi
.in -2
.sp

.sp
.LP
The preceding entries allow two jobs to run. The first one would run at
midnight in the GMT timezone and the second would run at midnight in the PST
timezone. Both would be run in the directory \fB/local/home/user\fR using the
Korn shell. The file concludes with \fBTZ\fR, \fBHOME\fR, and \fBSHELL\fR
entries that return those variable to their default values.

.SH ENVIRONMENT VARIABLES
.LP
See \fBenviron\fR(5) for descriptions of the following environment variables
that affect the execution of \fBcrontab\fR: \fBLANG\fR, \fBLC_ALL\fR,
\fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, and \fBNLSPATH\fR.
.SS "\fB/usr/bin/crontab\fR"
.ne 2
.na
\fB\fBEDITOR\fR\fR
.ad
.RS 10n
Determine the editor to be invoked when the \fB-e\fR option is specified. This
is overridden by the \fBVISUAL\fR environmental variable. The default editor is
\fBvi\fR(1).
.RE

.sp
.ne 2
.na
\fB\fBPATH\fR\fR
.ad
.RS 10n
The \fBPATH\fR in \fBcrontab\fR's environment specifies the search path used to
find the editor.
.RE

.sp
.ne 2
.na
\fB\fBVISUAL\fR\fR
.ad
.RS 10n
Determine the visual editor to be invoked when the \fB-e\fR option is
specified. If \fBVISUAL\fR is not specified, then the environment variable
\fBEDITOR\fR is used. If that is not set, the default is \fBvi\fR(1).
.RE

.SS "\fB/usr/xpg4/bin/crontab\fR"
.ne 2
.na
\fB\fBEDITOR\fR\fR
.ad
.RS 10n
Determine the editor to be invoked when the \fB-e\fR option is specified. The
default editor is \fB/usr/xpg4/bin/vi\fR.
.RE

.SS "\fB/usr/xpg6/bin/crontab\fR"
.ne 2
.na
\fB\fBEDITOR\fR\fR
.ad
.RS 10n
Determine the editor to be invoked when the \fB-e\fR option is specified. The
default editor is \fB/usr/xpg6/bin/vi\fR.
.RE

.SH EXIT STATUS
.LP
The following exit values are returned:
.sp
.ne 2
.na
\fB\fB0\fR\fR
.ad
.RS 6n
Successful completion.
.RE

.sp
.ne 2
.na
\fB\fB>0\fR\fR
.ad
.RS 6n
An error occurred.
.RE

.SH FILES
.ne 2
.na
\fB\fB/etc/cron.d\fR\fR
.ad
.RS 28n

usr/src/cmd/cron/crontab.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved	*/

/*
 * Copyright 2019 OmniOS Community Edition (OmniOSce) Association.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <ctype.h>
#include <pwd.h>
#include <unistd.h>
#include <locale.h>
#include <nl_types.h>
#include <langinfo.h>
#include <libintl.h>
#include <security/pam_appl.h>
#include <limits.h>
#include <libzoneinfo.h>
#include "cron.h"
#include "getresponse.h"

#if defined(XPG4)
#define	VIPATH	"/usr/xpg4/bin/vi"
#elif defined(XPG6)
#define	VIPATH	"/usr/xpg6/bin/vi"
#else
#define	_XPG_NOTDEFINED
#define	VIPATH	"vi"
#endif

#define	TMPFILE		"_cron"		/* prefix for tmp file */
#define	CRMODE		0600	/* mode for creating crontabs */

#define	BADCREATE	\
	"can't create your crontab file in the crontab directory."
#define	BADOPEN		"can't open your crontab file."
#define	BADSHELL	\
	"because your login shell isn't /usr/bin/sh, you can't use cron."
#define	WARNSHELL	"warning: commands will be executed using /usr/bin/sh\n"
#define	BADUSAGE	\
	"usage:\n"			\
	"\tcrontab [file]\n"		\
	"\tcrontab -e [username]\n"	\
	"\tcrontab -l [username]\n"	\
	"\tcrontab -r [username]"
#define	INVALIDUSER	"you are not a valid user (no entry in /etc/passwd)."
#define	NOTALLOWED	"you are not authorized to use cron.  Sorry."
#define	NOTROOT		\
	"you must be super-user to access another user's crontab file"
#define	AUDITREJECT	"The audit context for your shell has not been set."
#define	EOLN		"unexpected end of line."
#define	UNEXPECT	"unexpected character found in line."
#define	OUTOFBOUND	"number out of bounds."
#define OVERFLOW	"too many elements."
#define	ERRSFND		"errors detected in input, no crontab file generated."
#define	ED_ERROR	\
	"     The editor indicates that an error occurred while you were\n"\
	"     editing the crontab data - usually a minor typing error.\n\n"
#define	BADREAD		"error reading your crontab file"
#define	ED_PROMPT	\
	"     Edit again, to ensure crontab information is intact (%s/%s)?\n"\
	"     ('%s' will discard edits.)"
#define	NAMETOOLONG	"login name too long"
#define	BAD_TZ	"Timezone unrecognized in: %s"
#define	BAD_SHELL	"Invalid shell specified: %s"
#define	BAD_HOME	"Unable to access directory: %s\t%s\n"

extern int	per_errno;

extern int	audit_crontab_modify(char *, char *, int);
extern int	audit_crontab_delete(char *, int);
extern int	audit_crontab_not_allowed(uid_t, char *);

int		err;
int		cursor;
char		*cf;
char		*tnam;
char		edtemp[5+13+1];
char		line[CTLINESIZE];
static		char	login[UNAMESIZE];

static void	catch(int);
static void	crabort(char *);
static void	cerror(char *);
static void	copycron(FILE *);

int
main(int argc, char **argv)
{
	int	c, r;
	int	rflag	= 0;
	int	lflag	= 0;
	int	eflag	= 0;
	int	errflg	= 0;
	char *pp;
	FILE *fp, *tmpfp;
	struct stat stbuf;
	struct passwd *pwp;
	time_t omodtime;
	char *editor;
	uid_t ruid;
	pid_t pid;
	int stat_loc;
	int ret;
	char real_login[UNAMESIZE];
	int tmpfd = -1;
	pam_handle_t *pamh;
	int pam_error;
	char *buf;
	size_t buflen;

	(void) setlocale(LC_ALL, "");
#if !defined(TEXT_DOMAIN)	/* Should be defined by cc -D */
#define	TEXT_DOMAIN "SYS_TEST"	/* Use this only if it weren't */
#endif
	(void) textdomain(TEXT_DOMAIN);

	if (init_yes() < 0) {
		(void) fprintf(stderr, gettext(ERR_MSG_INIT_YES),
		    strerror(errno));
		exit(1);
	}

	while ((c = getopt(argc, argv, "elr")) != EOF)
		switch (c) {
			case 'e':
				eflag++;
				break;
			case 'l':
				lflag++;
				break;
			case 'r':
				rflag++;
				break;
			case '?':
				errflg++;
				break;
		}

	if (eflag + lflag + rflag > 1)
		errflg++;

	argc -= optind;
	argv += optind;
	if (errflg || argc > 1)
		crabort(BADUSAGE);

	ruid = getuid();
	if ((pwp = getpwuid(ruid)) == NULL)
		crabort(INVALIDUSER);

	if (strlcpy(real_login, pwp->pw_name, sizeof (real_login))
	    >= sizeof (real_login))
		crabort(NAMETOOLONG);

	if ((eflag || lflag || rflag) && argc == 1) {
		if ((pwp = getpwnam(*argv)) == NULL)
			crabort(INVALIDUSER);

		if (!cron_admin(real_login)) {
			if (pwp->pw_uid != ruid)
				crabort(NOTROOT);
			else
				pp = getuser(ruid);
		} else
			pp = *argv++;
	} else {
		pp = getuser(ruid);
	}

	if (pp == NULL) {
		if (per_errno == 2)
			crabort(BADSHELL);
		else
			crabort(INVALIDUSER);
	}
	if (strlcpy(login, pp, sizeof (login)) >= sizeof (login))
		crabort(NAMETOOLONG);
	if (!allowed(login, CRONALLOW, CRONDENY))
		crabort(NOTALLOWED);

	/* Do account validation check */
	pam_error = pam_start("cron", pp, NULL, &pamh);
	if (pam_error != PAM_SUCCESS) {
		crabort((char *)pam_strerror(pamh, pam_error));
	}
	pam_error = pam_acct_mgmt(pamh, PAM_SILENT);
	if (pam_error != PAM_SUCCESS) {
		(void) fprintf(stderr, gettext("Warning - Invalid account: "
		    "'%s' not allowed to execute cronjobs\n"), pp);
	}
	(void) pam_end(pamh, PAM_SUCCESS);


	/* check for unaudited shell */
	if (audit_crontab_not_allowed(ruid, pp))
		crabort(AUDITREJECT);

	cf = xmalloc(strlen(CRONDIR)+strlen(login)+2);
	strcat(strcat(strcpy(cf, CRONDIR), "/"), login);

	if (rflag) {
		r = unlink(cf);
		cron_sendmsg(DELETE, login, login, CRON);
		audit_crontab_delete(cf, r);
		exit(0);
	}
	if (lflag) {
		if ((fp = fopen(cf, "r")) == NULL)
			crabort(BADOPEN);
		while (fgets(line, CTLINESIZE, fp) != NULL)
			fputs(line, stdout);
		fclose(fp);
		exit(0);
	}
	if (eflag) {
		if ((fp = fopen(cf, "r")) == NULL) {
			if (errno != ENOENT)
				crabort(BADOPEN);
		}
		(void) strcpy(edtemp, "/tmp/crontabXXXXXX");
		tmpfd = mkstemp(edtemp);
		if (fchown(tmpfd, ruid, -1) == -1) {
			(void) close(tmpfd);
			crabort("fchown of temporary file failed");
		}
		(void) close(tmpfd);
		/*
		 * Fork off a child with user's permissions,
		 * to edit the crontab file
		 */
		if ((pid = fork()) == (pid_t)-1)
			crabort("fork failed");
		if (pid == 0) {		/* child process */
			/* give up super-user privileges. */
			setuid(ruid);
			if ((tmpfp = fopen(edtemp, "w")) == NULL)
				crabort("can't create temporary file");
			if (fp != NULL) {
				/*
				 * Copy user's crontab file to temporary file.
				 */
				while (fgets(line, CTLINESIZE, fp) != NULL) {
					fputs(line, tmpfp);
					if (ferror(tmpfp)) {
						fclose(fp);
						fclose(tmpfp);
						crabort("write error on"
						    "temporary file");
					}
				}
				if (ferror(fp)) {
					fclose(fp);
					fclose(tmpfp);
					crabort(BADREAD);
				}
				fclose(fp);
			}
			if (fclose(tmpfp) == EOF)
				crabort("write error on temporary file");
			if (stat(edtemp, &stbuf) < 0)
				crabort("can't stat temporary file");
			omodtime = stbuf.st_mtime;
#ifdef _XPG_NOTDEFINED
			editor = getenv("VISUAL");
			if (editor == NULL) {
#endif
				editor = getenv("EDITOR");
				if (editor == NULL)
					editor = VIPATH;
#ifdef _XPG_NOTDEFINED
			}
#endif
			buflen = strlen(editor) + strlen(edtemp) + 2;
			buf = xmalloc(buflen);
			(void) snprintf(buf, buflen, "%s %s", editor, edtemp);

			sleep(1);

			while (1) {
				ret = system(buf);

				/* sanity checks */
				if ((tmpfp = fopen(edtemp, "r")) == NULL)
					crabort("can't open temporary file");
				if (fstat(fileno(tmpfp), &stbuf) < 0)
					crabort("can't stat temporary file");
				if (stbuf.st_size == 0)
					crabort("temporary file empty");
				if (omodtime == stbuf.st_mtime) {
					(void) unlink(edtemp);
					fprintf(stderr, gettext(
					    "The crontab file was not"
					    " changed.\n"));
					exit(1);
				}
				if ((ret) && (errno != EINTR)) {
					/*
					 * Some editors (like 'vi') can return
					 * a non-zero exit status even though
					 * everything is okay. Need to check.
					 */
					fprintf(stderr, gettext(ED_ERROR));
					fflush(stderr);
					if (isatty(fileno(stdin))) {
						/* Interactive */
						fprintf(stdout,
						    gettext(ED_PROMPT),
						    yesstr, nostr, nostr);
						fflush(stdout);

						if (yes()) {
							/* Edit again */
							continue;
						} else {
							/* Dump changes */
							(void) unlink(edtemp);
							exit(1);
						}
					} else {
						/*
						 * Non-interactive, dump changes
						 */
						(void) unlink(edtemp);
						exit(1);
					}
				}
				exit(0);
			} /* while (1) */
		}

		/* fix for 1125555 - ignore common signals while waiting */
		(void) signal(SIGINT, SIG_IGN);
		(void) signal(SIGHUP, SIG_IGN);
		(void) signal(SIGQUIT, SIG_IGN);
		(void) signal(SIGTERM, SIG_IGN);
		wait(&stat_loc);
		if ((stat_loc & 0xFF00) != 0)
			exit(1);

		/*
		 * unlink edtemp as 'ruid'. The file contents will be held
		 * since we open the file descriptor 'tmpfp' before calling
		 * unlink.
		 */
		if (((ret = seteuid(ruid)) < 0) ||
		    ((tmpfp = fopen(edtemp, "r")) == NULL) ||
		    (unlink(edtemp) == -1)) {
			fprintf(stderr, "crontab: %s: %s\n",
			    edtemp, errmsg(errno));
			if ((ret < 0) || (tmpfp == NULL))
				(void) unlink(edtemp);
			exit(1);
		} else
			seteuid(0);

		copycron(tmpfp);
	} else {
		if (argc == 0)
			copycron(stdin);
		else if (seteuid(getuid()) != 0 || (fp = fopen(argv[0], "r"))
		    == NULL)
			crabort(BADOPEN);
		else {
			seteuid(0);
			copycron(fp);
		}
	}
	cron_sendmsg(ADD, login, login, CRON);
/*
 *	if (per_errno == 2)
 *		fprintf(stderr, gettext(WARNSHELL));
 */
	return (0);
}

static void
copycron(FILE *fp)
{
	FILE *tfp;
	char pid[6], *tnam_end;
	int t;
	char buf[LINE_MAX];
	cfield_error_t cferr;

	sprintf(pid, "%-5d", getpid());
	tnam = xmalloc(strlen(CRONDIR)+strlen(TMPFILE)+7);
	strcat(strcat(strcat(strcpy(tnam, CRONDIR), "/"), TMPFILE), pid);
	/* cut trailing blanks */
	tnam_end = strchr(tnam, ' ');
	if (tnam_end != NULL)
		*tnam_end = 0;
	/* catch SIGINT, SIGHUP, SIGQUIT signals */
	if (signal(SIGINT, catch) == SIG_IGN)
		signal(SIGINT, SIG_IGN);
	if (signal(SIGHUP, catch) == SIG_IGN) signal(SIGHUP, SIG_IGN);
	if (signal(SIGQUIT, catch) == SIG_IGN) signal(SIGQUIT, SIG_IGN);
	if (signal(SIGTERM, catch) == SIG_IGN) signal(SIGTERM, SIG_IGN);
	if ((t = creat(tnam, CRMODE)) == -1) crabort(BADCREATE);
	if ((tfp = fdopen(t, "w")) == NULL) {
		unlink(tnam);
		crabort(BADCREATE);
	}
	err = 0;	/* if errors found, err set to 1 */
	while (fgets(line, CTLINESIZE, fp) != NULL) {
		cursor = 0;
		while (line[cursor] == ' ' || line[cursor] == '\t')
			cursor++;
		/* fix for 1039689 - treat blank line like a comment */
		if (line[cursor] == '#' || line[cursor] == '\n')
			goto cont;

		if (strncmp(&line[cursor], ENV_TZ, strlen(ENV_TZ)) == 0) {
			char *x;

			strncpy(buf, &line[cursor + strlen(ENV_TZ)],
			    sizeof (buf));
			if ((x = strchr(buf, '\n')) != NULL)
				*x = '\0';

			if (isvalid_tz(buf, NULL, _VTZ_ALL)) {
				goto cont;
			} else {
				err = 1;
				fprintf(stderr, BAD_TZ, &line[cursor]);
				continue;
			}
		} else if (strncmp(&line[cursor], ENV_SHELL,
		    strlen(ENV_SHELL)) == 0) {
			char *x;

			strncpy(buf, &line[cursor + strlen(ENV_SHELL)],
			    sizeof (buf));
			if ((x = strchr(buf, '\n')) != NULL)
				*x = '\0';

			if (isvalid_shell(buf)) {
				goto cont;
			} else {
				err = 1;
				fprintf(stderr, BAD_SHELL, &line[cursor]);
				continue;
			}
		} else if (strncmp(&line[cursor], ENV_HOME,
		    strlen(ENV_HOME)) == 0) {
			char *x;

			strncpy(buf, &line[cursor + strlen(ENV_HOME)],
			    sizeof (buf));
			if ((x = strchr(buf, '\n')) != NULL)
				*x = '\0';
			if (chdir(buf) == 0) {
				goto cont;
			} else {
				err = 1;
				fprintf(stderr, BAD_HOME, &line[cursor],
				    strerror(errno));
				continue;
			}
		}

		if (next_field(0, 59, line, &cursor, &cferr) == NULL ||
		    next_field(0, 23, line, &cursor, &cferr) == NULL ||
		    next_field(1, 31, line, &cursor, &cferr) == NULL ||
		    next_field(1, 12, line, &cursor, &cferr) == NULL ||
		    next_field(0, 6, line, &cursor, &cferr) == NULL) {
			switch (cferr) {
			case CFIELD_EOLN:
				cerror(EOLN);
				break;
			case CFIELD_UNEXPECT:
				cerror(UNEXPECT);
				break;
			case CFIELD_OUTOFBOUND:
				cerror(OUTOFBOUND);
				break;
			case CFIELD_EOVERFLOW:
				cerror(OVERFLOW);
				break;
			case CFIELD_ENOMEM:
				(void) fprintf(stderr, "Out of memory\n");
				exit(55);
				break;
			default:
				break;
			}
			continue;
		}

		if (line[++cursor] == '\0') {
			cerror(EOLN);
			continue;
		}
cont:
		if (fputs(line, tfp) == EOF) {
			unlink(tnam);
			crabort(BADCREATE);
		}
	}
	fclose(fp);
	fclose(tfp);

	/* audit differences between old and new crontabs */
	audit_crontab_modify(cf, tnam, err);

	if (!err) {
		/* make file tfp the new crontab */
		unlink(cf);
		if (link(tnam, cf) == -1) {
			unlink(tnam);
			crabort(BADCREATE);
		}
	} else {
		crabort(ERRSFND);
	}
	unlink(tnam);
}

static void
cerror(char *msg)
{
	fprintf(stderr, gettext("%scrontab: error on previous line; %s\n"),
	    line, msg);
	err = 1;
}


static void
catch(int x)
{
	unlink(tnam);
	exit(1);
}

static void
crabort(char *msg)
{
	int sverrno;

	if (strcmp(edtemp, "") != 0) {
		sverrno = errno;
		(void) unlink(edtemp);
		errno = sverrno;
	}
	if (tnam != NULL) {
		sverrno = errno;
		(void) unlink(tnam);
		errno = sverrno;
	}
	fprintf(stderr, "crontab: %s\n", gettext(msg));
	exit(1);
}
